src/stdlib/mod.rs

```diff
@@ -163,3 +169,127 @@ fn assert<B: Backend>(
 
     Ok(None)
 }
+
+fn to_bits<B: Backend>(
+    compiler: &mut CircuitWriter<B>,
+    generics: &GenericParameters,
+    vars: &[VarInfo<B::Field, B::Var>],
+    span: Span,
+) -> Result<Option<Var<B::Field, B::Var>>> {
+    // should be two input vars
+    assert_eq!(vars.len(), 2);
+
+    // but the better practice would be to retrieve the value from the generics
+    let num_var = generics.get("LEN");
+
+    // num should be greater than 0
+    assert!(num_var > 0);
+
+    // alternatively, it can be retrieved from the first var, but it is not recommended
+    // let num_var = &vars[0];
+
+    let var_info = &vars[1];
+    let var = &var_info.var;
+    assert_eq!(var.len(), 1);
+
+    // second var is the value to convert
+    let val = match &var[0] {
+        ConstOrCell::Cell(cvar) => cvar,
+        ConstOrCell::Const(_) => todo!(),
+    };
+
+    // convert value to bits
+    let mut bits = Vec::new();
+    let mut e2 = 1u32;
+    let one = B::Field::one();
+    let zero = B::Field::zero();
+    let mut lc: Option<B::Var> = None;
+
+    // this constant is to adapt to the backend::sub api which only accepts vars as operands
+    let var_one = compiler
+        .backend
+        .add_constant(Some("constant one"), one, span);
+
+    for i in 0..num_var {
+        // value >> i & 1
+        let bit = compiler
+            .backend
+            .new_internal_var(Value::RSAnd(val.clone(), i as usize), Span::new(2, 1, 1));
+
+        // constrain it to be either 0 or 1
+        // bits[i] * (bits[i] - 1 ) === 0;
+        // todo: might need to add sub_const to the backend?
+        let bit_minus_one = compiler.backend.sub(&bit, &var_one, Span::new(2, 2, 1));
+        let zero_mul = compiler
+            .backend
+            .mul(&bit, &bit_minus_one, Span::new(2, 3, 1));
+        compiler
+            .backend
+            .assert_eq_const(&zero_mul, zero, Span::new(2, 4, 1));
+
+        // lc += bits[i] * e2;
+        let weighted_bit =
+            compiler
+                .backend
+                .mul_const(&bit, &B::Field::from(e2), Span::new(2, 5, 1));
```

Comment:
why these fake spans btw :D?

Reply:
It was for debugging. I should remove them.